src/MonoidalCategory/FreeMonoidalCategory.lidr

```diff
@@ -0,0 +1,273 @@
+\iffalse
+SPDX-License-Identifier: AGPL-3.0-only
+
+This file is part of `idris-ct` Category Theory in Idris library.
+
+Copyright (C) 2019 Stichting Statebox <https://statebox.nl>
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU Affero General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program.  If not, see <https://www.gnu.org/licenses/>.
+\fi
+
+> module FreeMonoidalCategory
+>
+> import Basic.Category
+> import Basic.Functor
+> import Basic.Isomorphism
+> import Basic.NaturalIsomorphism
+> import Basic.NaturalTransformation
+> import Data.List
+> import Monoid.FreeMonoid
+> import Monoid.Monoid
+> import MonoidalCategory.StrictMonoidalCategory
+> import MonoidalCategory.StrictSymmetricMonoidalCategory
+> import MonoidalCategory.SymmetricMonoidalCategoryHelpers
+> import Product.ProductCategory
+>
+> %access public export
+> %default total
+>
+> data FreeMorphism :
+>      (t : Type)
+>   -> (generatingMorphisms : List (List t, List t))
+>   -> (domain : List t)
+>   -> (codomain : List t)
+>   -> Type
+> where
+>   MkIdFreeMorphism            : (x : List t) -> FreeMorphism t generatingMorphisms x x
+>   MkSymmetryFreeMorphism      : (x, y : List t) -> FreeMorphism t generatingMorphisms (x ++ y) (y ++ x)
+>   MkCompositionFreeMorphism   : FreeMorphism t generatingMorphisms a b
+>                              -> FreeMorphism t generatingMorphisms b c
+>                              -> FreeMorphism t generatingMorphisms a c
+>   MkJuxtapositionFreeMorphism : FreeMorphism t generatingMorphisms a b
+>                              -> FreeMorphism t generatingMorphisms c d
+>                              -> FreeMorphism t generatingMorphisms (a ++ c) (b ++ d)
+>   MkGeneratingFreeMorphism    : (l : List (List t, List t))
+>                              -> (e : (List t, List t))
+>                              -> Elem e l
+>                              -> FreeMorphism t l (Basics.fst e) (Basics.snd e)
+>
+> freeIdentity : (ts : List t) -> FreeMorphism t generatingMorphisms ts ts
+> freeIdentity = MkIdFreeMorphism
+>
+> freeComposition :
+>      (as, bs, cs : List t)
+>   -> (fm1 : FreeMorphism t generatingMorphisms as bs)
+>   -> (fm2 : FreeMorphism t generatingMorphisms bs cs)
+>   -> FreeMorphism t generatingMorphisms as cs
+> freeComposition as bs cs fm1 fm2 = MkCompositionFreeMorphism fm1 fm2
+>
+> postulate
+> freeLeftIdentity :
```

Comment:
These postulates makes it possible to prove false:
```idris
oops : Void
oops = constructorsDisjoint (freeLeftIdentity [] [] (MkIdFreeMorphism []))
  where
    DistinguishThem : FreeMorphism () [] [] [] -> Type
    DistinguishThem (MkIdFreeMorphism []) = Void
    DistinguishThem _                     = ()
    constructorsDisjoint : MkCompositionFreeMorphism {t = ()} {generatingMorphisms = []} (MkIdFreeMorphism []) (MkIdFreeMorphism []) = (MkIdFreeMorphism []) -> Void
    constructorsDisjoint p = replace {P = DistinguishThem} p ()
```

What you need to pursue this approach is some form of [quotient type](https://ncatlab.org/nlab/show/higher+inductive+type#QuotientsOfSets), which Idris unfortunately doesn't support.

But maybe there is some clever way to construct the free strict monoidal category explicitly, e.g. the free category on a graph can be constructed as the transitive closure of the edge relation (in other words, a morphism is a list of edges), which gives you the category laws for free, instead of having to quotient by them explicitly.

Reply:
This is fucked :( Do you have a reference for building free cats over graphs? Are you referring to the same construction one can find, say, on MacLane?

Also, should it be possible to generalize to monoidal categories by using hypegraphs as generators? :)

Reply:
I don't have access to MacLane at the moment, but it's probably the same construction. The nlab is quite explicit, and has further references:
https://ncatlab.org/nlab/show/path+category

From a type theory point of view, what is nice with this construction is that it is only involving free concepts (that hence can be implemented by inductive definitions), with no need to quotient. Compare with the construction of the free monoid, which is a very similar situation. I'll have to think about if hypergraphs would work. My immediate concern would be that we now also have to introduce morphisms not only from compositions of previously introduced ones, but also by applying the functoriality of the tensor, which I fear would need something more than just paths of edges.

Reply:
I've dealt with similar issues a lot in Coq and Agda. I don't know if the path category approach works, but there are several ways to deal with the issue that quotient types don't exist.

In this scenario, I would postulate the existence of the quotient type locally. I.e. keep the type `FreeMorphism` as it is but rename it to `PreFreeMorphism`, and then postulate the existence of a new type `FreeMorphism` together with a map `quot : PreFreeMorphism ... -> FreeMorphism ...`, such that the type `FreeMorphism` has exactly the structure that we want (i.e. the map is surjective, and `quot x = quot y` if and only if they are of the forms as described by the postulates here). This is for example how real numbers are implemented in Coq.

Alternatively, one could deal with setoids, but everytime I did I regretted it big time.

Reply:
My gut feeling is that trying to go with what Fred said, but using hypegraphs, is probably worth trying first. If this experience has taught me something it is that postulate is a very dangerous construct to use. So if there's a way of giving definitions which avoids it completely, I'd gladly go for it.

In addition to it, Petri nets can clearly be seen as hypegraphs, where places are vertexes and transitions are hyperedges. Generating a free smc from an hypegraph should basically be exactly what we want (you have to add symmetries but I don't want to think about this right now :) )

Reply:
@fredrikNordvallForsberg I fiddled with this a bit and I don't think the hypegraph approach can work. Moreover, strictification makes it even worse. I frankly don't have any idea about how to proceed, any help is appreciated.

Reply:
could something like quotient inductive types (https://crunchytype.wordpress.com/2017/04/27/quotient-inductive-types-their-dual/) work?

Reply:
Thinking about it again, if there was a convenient way to store hypergraphs then free monoidal categories with hypegraphs may be doable (dunno about symmetries tho). 
Specifically: 
- Objects are ordered lists. 
- Morphisms are directed hypegraphs such that:
  -  the objects in source/target list form a subset of its vertexes;
  -  objects in the source list are vertexes with exactly one outbound hyperedge
  -  objects in target list are vertexes with exactly one inbound hyperedge.

In this setting identities are represented as hypegraphs with no hyperedges. Composition is just gluing hypegraphs along common edges, while tensor product is just list concatenation on objects, while stacking hypegraphs one under the other on morphisms.

All this is modulo some details that still have to be checked. Anyway, in this setting, we could generate a free monoidal category by providing a list of object generators (and doing the list monoid on them) and a list of ¨basic hypegraphs¨ that have object generators as vertexes.
The morphisms are given by all the possible gluings of hypegraphs with each other.

Reply:
looking around I found this (https://homotopytypetheory.org/2011/04/23/running-circles-around-in-your-proof-assistant/) which could help us save what we did. We could use something along these lines:

```
module Test

%access public export
%default total

private
data IHidden : Type where
  Zero : IHidden
  One  : IHidden

export
I : Type
I = IHidden

export
zero : I
zero = Zero

export
one : I
one = One

postulate
zeroEqualsOne : zero = one

Ielimination : (x, y : a) -> (p : x = y) -> I -> a
Ielimination x y p Zero = x
Ielimination x y p Zero = y
```

so that it becomes impossible to define something like `DistinguishThem` as @fredrikNordvallForsberg did above outside this `Test` module

Reply:
Isn't this something similar to what @WhatisRT suggested?

Reply:
yes, I guess they are practically the same approach. This is just a concrete implementation

Reply:
Well then we know that at least two people in the team like this now!
Anyway, if this fixes it, go for it. Nevertheless, I'd consider it more of a momentary fix that a definitive thing: If we can get stuff done in terms of free compositions of other structures I'd feel way better :)

Reply:
In principle, other than using setoids this is the only solution in Idris, so if you consider this to be only a temporary fix, we either have to make large changes (like switching to a language that supports quotients or using setoids everywhere), or we have to give a quotient type explicitly.

In this scenario, there might actually be a reasonable way to do this. As long as the sets of generators are finite, the set of objects should be countable infinite and the set of morphisms between two objects should be finite (I haven't given much thought to it, maybe it is possible that some are coubtably infinite?). Thus, it should be possible to figure out the size of all the required types, and you would be able to construct something that is isomorphic to the quotient you want, without using postulates. That would be much more work though.

As an alternative, Agda 2.6 supports cubical type theory, so in particular quotient inductive types. Translating Idris code to Agda is reasonably quick and I have already translated everything in the `Basic` directory. So we could just postulate that the quotients exist in a different file, and then have an Agda copy of that file where we just add the definition of the quotient we want and replace the postulates by theorems there. If it's just a definition of that one quotient, it's unlikely that that will change, so it shouldn't be a maintenance problem.

Reply:
Did you read my previous comments in terms of free structures? Why do you think it is not possible to do anything like this? It works for the path category, do you have elements to affirm there is no possible construction of FSSMCS in terms of free structures?

Reply:
To recap what we've decided in the meeting: It seems best to use postulates right now, to just give us the existence of the type we need. I think the proposal by @marcosh is a great start, but it should be done in such a way that the implementation of FSSMCS is parametrized over the implementation of the quotient. This would mean that whenever we find a good way to construct the type we need without postulates, we can just change a single line of code to switch between the implementation with postulates and any other implementation. Afterwards, we can decide how to tackle the implementation of that type, either via a subtype using normal forms, doing some random stuff with isomorphisms or by getting some of @FabrizioRomanoGenovese's ideas to work.